parachain-system/src/lib.rs

```diff
@@ -443,6 +432,126 @@ decl_module! {
+			Error::<T>::DmpMqcMismatch
+		);
+
+		// Store the processed_downward_messages here so that it will be accessible from
+		// PVF's `validate_block` wrapper and collation pipeline.
+		storage::unhashed::put(well_known_keys::PROCESSED_DOWNWARD_MESSAGES, &dm_count);
+
+		Ok(())
+	}
+
+	/// Process all inbound horizontal messages relayed by the collator.
+	///
+	/// This is similar to [`process_inbound_downward_messages`], but works on multiple inbound
+	/// channels.
+	fn process_inbound_horizontal_messages(
+		vfp: &PersistedValidationData,
+		horizontal_messages: BTreeMap<ParaId, Vec<InboundHrmpMessage>>,
+	) -> DispatchResult {
+		// First, check that all submitted messages are sent from channels that exist. The channel
+		// exists if its MQC head is present in `vfp.hrmp_mqc_heads`.
+		for sender in horizontal_messages.keys() {
+			ensure!(
+				vfp.hrmp_mqc_heads
+					.binary_search_by_key(sender, |&(s, _)| s)
+					.is_ok(),
+				Error::<T>::HrmpNoMqc,
+			);
+		}
+
+		// Second, prepare horizontal messages for a more convenient processing:
+		//
+		// instead of a mapping from a para to a list of inbound HRMP messages, we will have a list
+		// of tuples `(sender, message)` first ordered by `sent_at` (the relay chain block number
+		// in which the message hit the relay-chain) and second ordered by para id ascending.
+		//
+		// The messages will be dispatched in this order.
+		let mut horizontal_messages = horizontal_messages
+			.into_iter()
+			.flat_map(|(sender, channel_contents)| {
+				channel_contents
+					.into_iter()
+					.map(move |message| (sender, message))
+			})
+			.collect::<Vec<_>>();
+		horizontal_messages.sort_by(|a, b| {
+			// first sort by sent-at and then by the para id
+			match a.1.sent_at.cmp(&b.1.sent_at) {
+				cmp::Ordering::Equal => a.0.cmp(&b.0),
+				ord => ord,
+			}
+		});
+
+		let last_mqc_heads = LastHrmpMqcHeads::get();
+		let mut running_mqc_heads = BTreeMap::new();
+		let mut hrmp_watermark = None;
+
+		for (sender, horizontal_message) in horizontal_messages {
+			if hrmp_watermark
+				.map(|w| w < horizontal_message.sent_at)
+				.unwrap_or(true)
+			{
+				hrmp_watermark = Some(horizontal_message.sent_at);
+			}
+
+			running_mqc_heads
+				.entry(sender)
+				.or_insert_with(|| last_mqc_heads.get(&sender).cloned().unwrap_or_default())
+				.extend_hrmp(&horizontal_message);
+
+			T::HrmpMessageHandlers::handle_hrmp_message(sender, horizontal_message);
+		}
+
+		// Check that the MQC heads for each channel provided by the relay chain match the MQC heads
+		// we have after processing all incoming messages.
+		//
+		// Along the way we also carry over the relevant entries from the `last_mqc_heads` to
+		// `running_mqc_heads`. Otherwise, in a block where no messages were sent in a channel
+		// it won't get into next block's `last_mqc_heads` and thus will be all zeros, which
+		// would corrupt the message queue chain.
+		for &(ref sender, ref target_head) in &vfp.hrmp_mqc_heads {
```

Comment:
I assume these are always all heads, even from the paras that didn't send us any messages in this block? :D

Reply:
Yeah, those are all. Good idea to leave a comment about that on `hrmp_mqc_heads`.

Reply:
Ah, wait, I just removed it in the [next PR](https://github.com/paritytech/cumulus/pull/317) 🤦‍♂️